<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>

# Introduction  Notebook

Estimated time needed: **10** minutes

## Objectives

After completing this lab you will be able to:

*   Acquire data in various ways
*   Obtain insights from data with Pandas library


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="https://#data_acquisition">Data Acquisition</a>
    <li><a href="https://#basic_insight">Basic Insight of Dataset</a></li>
</ol>

</div>
<hr>


<h1 id="data_acquisition">Data Acquisition</h1>
<p>
There are various formats for a dataset: .csv, .json, .xlsx  etc. The dataset can be stored in different places, on your local machine or sometimes online.<br>

In this section, you will learn how to load a dataset into our Jupyter Notebook.<br>

In our case, the Trading Dataset is an online source, and it is in a CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.

<ul>
    <li>Data source: <a href="https://1824251045.rsc.cdn77.org/web/algohouse/data/BTCBUSD_trades_1m.csv">https://1824251045.rsc.cdn77.org/web/algohouse/data/BTCBUSD_trades_1m.csv</a></li>
    <li>Data type: csv</li>
</ul>
The Pandas Library is a useful tool that enables us to read various datasets into a dataframe; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


In [3]:
#you are running the lab in your  browser, so we will install the libraries using ``piplite``
import piplite
import micropip
await piplite.install(['pandas'])
await piplite.install(['matplotlib'])
await piplite.install(['scipy'])
await piplite.install(['seaborn'])
await micropip.install(['ipywidgets'],keep_going=True)
await micropip.install(['tqdm'],keep_going=True)

If you run the lab locally using Anaconda, you can load the correct library and versions by uncommenting the following:


In [4]:
#install specific version of libraries used in  lab
#! mamba install pandas==1.3.3  -y
#! mamba install numpy=1.21.2 -y

In [5]:
# import pandas library
import pandas as pd
import numpy as np

This function will download the dataset into your browser


In [6]:
#This function will download the dataset into your browser 

from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

<h2>Read Data</h2>
<p>
We use <code>pandas.read_csv()</code> function to read the csv file. In the brackets, we put the file path along with a quotation mark so that pandas will read the file into a dataframe from that address. The file path can be either an URL or your local file address.<br>

Because the data does not include headers, we can add an argument <code>headers = None</code> inside the <code>read_csv()</code> method so that pandas will not automatically set the first row as a header.<br>

You can also assign the dataset to any variable you create.

</p>


In [7]:
path = "https://1824251045.rsc.cdn77.org/web/algohouse/data/BTCBUSD_trades_1m.csv"

you will need to download the dataset; if you are running locally, please comment out the following


In [8]:
#you will need to download the dataset; if you are running locally, please comment out the following 
await download(path, "BTCBUSD_trades.csv")
path="BTCBUSD_trades.csv"

In [9]:
# Import pandas library
import pandas as pd

# Read the online file by the URL provides above, and assign it to variable "df", use index_col = 0 to make first column as indexes, 
# low_memory = False not to get errors with data types 

df = pd.read_csv(path, index_col=0, low_memory=False)


Further specify the value of the precision parameter equal to 2 to display two decimal signs (instead of 6 as default).

In [28]:
pd.options.display.float_format = '{:.2f}'.format

After reading the dataset, we can use the <code>dataframe.head(n)</code> method to check the top n rows of the dataframe, where n is an integer. Contrary to <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> will show you the bottom n rows of the dataframe.


In [29]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(5)

The first 5 rows of the dataframe


time     open     high      low    close  volume  rec_count  \
0 2022-11-11 14:38:00 16540.10 16540.10 16486.50 16486.50  304.95         97   
1 2022-11-11 14:39:00 16489.30 16493.30 16456.20 16462.10  319.26        103   
2 2022-11-11 14:40:00 16465.50 16496.00 16447.40 16492.90  290.90         99   
3 2022-11-11 14:41:00 16492.90 16547.10 16472.50 16543.00  314.05         99   
4 2022-11-11 14:42:00 16540.20 16547.10 16498.20 16498.30  226.95         99   

   avg_price  
0   16516.81  
1   16473.84  
2   16478.64  
3   16516.50  
4   16527.09

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #1: </h1>
<b>Check the bottom 10 rows of data frame "df".</b>
</div>


In [30]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
print("The last 10 rows of the dataframe\n")
df.tail(10)
```


<h3>Add Headers</h3>
<p>
Take a look at our dataset. Pandas automatically set the header with an integer starting from 0.
</p>
<p>
To better describe our data, we can introduce a header. This information is available at:  <a href="https://1824251045.rsc.cdn77.org/web/algohouse/data/BTCBUSD_trades_1m.csv" target="_blank">https://1824251045.rsc.cdn77.org/web/algohouse/data/BTCBUSD_trades_1m.csv</a>.
</p>
<p>
Thus, we have to add headers manually.
</p>
<p>
First, we create a list "headers" that include all column names in order.
Then, we use <code>dataframe.columns = headers</code> to replace the headers with the list we created.
</p>


In [31]:
# create headers list
headers = ["time","open","high","low","close","volume","rec_count","avg_price"]
print("headers\n", headers)

headers
 ['time', 'open', 'high', 'low', 'close', 'volume', 'rec_count', 'avg_price']


We replace headers and recheck our dataframe:


In [35]:
df.columns = headers
df.head(10)

time     open     high      low    close  volume  rec_count  \
0 2022-11-11 14:38:00 16540.10 16540.10 16486.50 16486.50  304.95         97   
1 2022-11-11 14:39:00 16489.30 16493.30 16456.20 16462.10  319.26        103   
2 2022-11-11 14:40:00 16465.50 16496.00 16447.40 16492.90  290.90         99   
3 2022-11-11 14:41:00 16492.90 16547.10 16472.50 16543.00  314.05         99   
4 2022-11-11 14:42:00 16540.20 16547.10 16498.20 16498.30  226.95         99   
5 2022-11-11 14:43:00 16498.30 16548.80 16491.90 16499.80  219.36         96   
6 2022-11-11 14:44:00 16502.30 16542.00 16502.30 16540.80  233.96        101   
7 2022-11-11 14:45:00 16540.70 16746.00 16540.50 16742.70 1183.99        112   
8 2022-11-11 14:46:00 16747.60 16829.40 16686.90 16829.40 1356.19        113   
9 2022-11-11 14:47:00 16829.80 17285.20 16828.30 17212.90 2122.80        116   

   avg_price  
0   16516.81  
1   16473.84  
2   16478.64  
3   16516.50  
4   16527.09  
5   16519.31  
6   16526.21  
7   16639.43  
8   16760.05  
9   17060.89

Now, we have successfully read the raw dataset and added the correct headers into the dataframe.


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>
<b>Find the name of the columns of the dataframe.</b>
</div>


In [36]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
print(df.columns)
```

</details>


<h2>Save Dataset</h2>
<p>
Correspondingly, Pandas enables us to save the dataset to csv. By using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
For example, if you would save the dataframe <b>df</b> as <b>btc.csv</b> to your local machine, you may use the syntax below, where <code>index = False</code> means the row names will not be written.
</p>


In [37]:
df.to_csv("btc.csv", index=False)

We can also read and save other file formats. We can use similar functions like **`pd.read_csv()`** and **`df.to_csv()`** for other data formats. The functions are listed in the following table:


<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


<h1 id="basic_insight">Basic Insight of Dataset</h1>
<p>
After reading data into Pandas dataframe, it is time for us to explore the dataset.<br>

There are several ways to obtain essential insights of the data to help us better understand our dataset.

</p>


<h2>Data Types</h2>
<p>
Data has a variety of types.<br>

The main types stored in Pandas dataframes are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:

</p>


In [38]:
df.dtypes


time         datetime64[ns]
open                float64
high                float64
low                 float64
close               float64
volume              float64
rec_count             int64
avg_price           float64
dtype: object

A series with the data type of each column is returned.


<p>
    As shown above, it is clear to see that the data type of "time" is <b>object</b>, "rec_count" - <b>int64</b>, etc.
</p>
<p>
These data types can be changed; we will learn how to accomplish this in a later module.
</p>


<h2>Describe</h2>
If we would like to get a statistical summary of each column e.g. count, column mean value, column standard deviation, etc., we use the describe method:


```python
dataframe.describe()
```

This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [39]:
df.describe()

open     high      low    close   volume  rec_count  avg_price
count 67291.00 67291.00 67291.00 67291.00 67291.00   67291.00   67291.00
mean  16793.17 16796.67 16789.78 16793.17    47.14      42.27   16793.26
std     370.66   369.95   371.38   370.67   107.36      21.76     370.62
min   15478.50 15511.60 15443.50 15475.10     0.01       1.00   15479.29
25%   16574.80 16578.80 16571.50 16574.90     9.29      25.00   16574.85
50%   16803.30 16806.30 16800.40 16803.30    22.63      38.00   16803.21
75%   16968.00 16970.90 16965.40 16968.00    48.15      58.00   16968.07
max   18344.60 18370.00 18324.20 18344.70  3986.89     124.00   18352.16

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute "open" has 67291 counts, the mean value of this column is 16793.167007, the standard deviation is 370.661830, the minimum value is 15478.5, 25th percentile is 16574.8, 50th percentile is 16803.3, 75th percentile is 16968, and the maximum value is 18344.6. <br>

However, what if we would also like to check all the columns including those that are of type object? <br><br>

You can add an argument <code>include = "all"</code> inside the bracket. Let's try it again.

</p>


In [40]:
# describe all the columns in "df", use datetime_is_numeric=True to make datetime64 == numeric
df.describe(include = "all",datetime_is_numeric=True)

time     open     high      low    close  \
count                          67291 67291.00 67291.00 67291.00 67291.00   
mean   2022-12-05 18:04:34.706573056 16793.17 16796.67 16789.78 16793.17   
min              2022-11-11 14:38:00 15478.50 15511.60 15443.50 15475.10   
25%              2022-11-24 03:08:30 16574.80 16578.80 16571.50 16574.90   
50%              2022-12-05 19:37:00 16803.30 16806.30 16800.40 16803.30   
75%              2022-12-17 12:38:30 16968.00 16970.90 16965.40 16968.00   
max              2022-12-29 23:59:00 18344.60 18370.00 18324.20 18344.70   
std                              NaN   370.66   369.95   371.38   370.67   

        volume  rec_count  avg_price  
count 67291.00   67291.00   67291.00  
mean     47.14      42.27   16793.26  
min       0.01       1.00   15479.29  
25%       9.29      25.00   16574.85  
50%      22.63      38.00   16803.21  
75%      48.15      58.00   16968.07  
max    3986.89     124.00   18352.16  
std     107.36      21.76     370.62

<p>
Now it provides the statistical summary of all the columns, including object-typed attributes.<br>

We can now see how many unique values there, which one is the top value and the frequency of top value in the object-typed columns.<br>

Some values in the table above show as "NaN". This is because those numbers are not available regarding a particular column type.<br>

</p>


Also we can describe only one column:

In [41]:
df[['time']].describe(datetime_is_numeric=True)

time
count                          67291
mean   2022-12-05 18:04:34.706573056
min              2022-11-11 14:38:00
25%              2022-11-24 03:08:30
50%              2022-12-05 19:37:00
75%              2022-12-17 12:38:30
max              2022-12-29 23:59:00

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #3: </h1>

<p>
You can select the columns of a dataframe by indicating the name of each column. For example, you can select the three columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Where "column" is the name of the column, you can apply the method  ".describe()" to get the statistics of those columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>

Apply the  method to ".describe()" to the columns 'length' and 'compression-ratio'.

</div>


In [42]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
df[['open', 'avg_price']].describe()
```

</details>


<h2>Info</h2>
Another method you can use to check your dataset is:


```python
dataframe.info()
```

It provides a concise summary of your DataFrame.

This method prints information about a DataFrame including the index dtype and columns, non-null values and memory usage.


In [43]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 67291 entries, 0 to 67290
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   time       67291 non-null  datetime64[ns]
 1   open       67291 non-null  float64       
 2   high       67291 non-null  float64       
 3   low        67291 non-null  float64       
 4   close      67291 non-null  float64       
 5   volume     67291 non-null  float64       
 6   rec_count  67291 non-null  int64         
 7   avg_price  67291 non-null  float64       
dtypes: datetime64[ns](1), float64(6), int64(1)
memory usage: 4.6 MB


<h1>Excellent! You have just completed the  Introduction Notebook!</h1>


### Thank you for completing this lab!

## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Joseph Santarcangelo</a>

### Other Contributors

<a href="https://www.linkedin.com/in/mahdi-noorian-58219234/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Mahdi Noorian PhD</a>

Bahare Talayian

Eric Xiao

Steven Dong

Parizad

Hima Vasudevan

<a href="https://www.linkedin.com/in/fiorellawever/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Fiorella Wenver</a>

<a href="https:// https://www.linkedin.com/in/yi-leng-yao-84451275/ " target="_blank" >Yi Yao</a>

<a href="https://www.linkedin.com/in/bohdan-tsisinskyi-539913255/ " target="_blank" >Bohdan Tsisinskyi</a>.


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                                         |
| ----------------- | ------- | ---------- | ---------------------------------------------------------- |
| 2023-01-18        | 3.0     | Bohdan     | Change dataset, add some parameters, comments
| 2022-08-23        | 2.4     | Malika     | Import micropip added and parameter for ipwidgets and tqdm |
| 2020-10-30        | 2.3     | Lakshmi    | Changed URL of the csv                                     |
| 2020-09-22        | 2.2     | Nayef      | Added replace() method to remove '?'                       |
| 2020-09-09        | 2.1     | Lakshmi    | Made changes in info method of dataframe                   |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab                         |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
